add contributing docs for i18n support (#5416)
foxglove/studio: CONTRIBUTING.md

1. Launch the development environment:

```sh
# To launch the desktop app (run both scripts concurrently):
$ yarn desktop:serve        # start webpack
$ yarn desktop:start        # launch electron

# To launch the browser app:
$ yarn web:serve

# To launch the browser app using a local instance of the backend server:
$ yarn web:serve:local

# To launch the storybook:
$ yarn storybook

# Advanced usage: running webpack and electron on different computers (or VMs) on the same network
$ yarn desktop:serve --host 192.168.xxx.yyy         # the address where electron can reach the webpack dev server
$ yarn dlx electron@22.1.0 .webpack # launch the version of electron for the current computer's platform

# To launch the desktop app using production API endpoints
$ yarn desktop:serve --env FOXGLOVE_BACKEND=production
$ yarn desktop:start

# NOTE: yarn web:serve does not support connecting to the production endpoints
```

### Other useful commands

```sh
$ yarn run          # list available commands
$ yarn lint         # lint all files
$ yarn test         # run all tests
$ yarn test:watch   # run tests on changed files
```

> ## Localization (multi-language support)

Reviewer: ```suggestion
## Localization
```

> Foxglove Studio primarily supports English, but a Chinese localization is available with translations provided by community volunteers. Translation support is implemented using [`react-i18next`](https://react.i18next.com).

Reviewer 2: ```suggestion
At this time, first class support for Foxglove Studio is English language only. Localization for other languages is available on a best-effort basis with translations provided by community volunteers. Translation support is implemented using [`react-i18next`](https://react.i18next.com). Current community supported languages available are:
* Chinese
```

Reviewer 2: Wanted to make it clear community support is best-effort and set it up in a way to easily expand the list.

> ### Guidelines for translations

Reviewer: ```suggestion
### Adding translations
```

> - English text **must** be kept up to date in every pull request. Updating translations for other languages in the same pull request is **optional**. We value _high-quality_ translations more than having _some_ translation for each part of the user interface.

Reviewer: Not sure what you mean by "We value _high-quality_ translations more than having _some_ translation for each part of the user interface." So if a user has an incomplete translation set, but it's high quality, we want them to contribute?

Author: I expect most PRs will fall into one of two groups. It's possible that some PRs would do both of these, but I'm just guessing:

1. adding or changing features and _only_ updating English (if these PRs break any existing Chinese text, it's better for them to remove the affected Chinese strings rather than have them become wrong)
2. not adding or changing UI features at all for English, only adding/updating Chinese translations

The point I was trying to get at is that for someone using the app in Chinese, it's better if all the Chinese text makes sense but some text is in (correct) English, rather than having it _all_ be in Chinese but some of the Chinese being mediocre/incorrect.

Reviewer: Got it. 

Maybe something like this then?

"We value having _high-quality_ translations over having _all_ translations for a given component or view. Though all English text **must** be kept up to date in every pull request, do not update other languages' translations in the same PR if an accurate translation is not yet available."

Author: ```suggestion
- We value having _high-quality_ translations over having _all_ translations for a given component or view. Though every PR must have up-to-date English translations, updating other languages is completely optional. We encourage you to use follow-up PRs to update other languages, once you have accurate translations.
```

> - If you update an English translation and cannot provide updated non-English translations, delete the non-English versions in that PR. Open follow-up PRs to handle adding accurate non-English translations.

Reviewer: Now that I see the edited list, the last sentence of each bullet point is basically saying the same thing.

Can we make consistent, pull out into a 3rd bullet point, or just use once where it makes more sense?

My preference:

- We value having _high-quality_ translations over having _all_ translations for a given component or view. Though every PR must have up-to-date English translations, updating other languages is completely optional.
- If you update an English translation and cannot provide updated non-English translations, delete the non-English versions in that PR. Open follow-up PRs to add accurate non-English translations.

Reviewer 3: "Open follow-up PRs to handle adding accurate non-English translations." Do I need to do that when I change the english translations? This makes it sound like I need to do that immediately? Or when?

Reviewer 2: ^^ Good point, don't want to make it sound required. Perhaps add "Optionally open follow-up..."?

> - If the English version of an existing translation is changed, and other versions would need to be updated, is better to delete the non-English versions rather than leave them unchanged, since they might become incorrect or confusing.

Reviewer: Wording here is also confusing. 

Not sure this is what you meant, but maybe `If an English translation is updated, delete the non-English versions in that PR. Open follow-up PRs to handle adding accurate non-English translations."?

Author: ```suggestion
- If you update an English translation and cannot provide updated non-English translations, delete the non-English versions in that PR. Open follow-up PRs to handle adding accurate non-English translations.
```

> ### Place strings in the `i18n` directory

Reviewer: ```suggestion
### Add translations to the `i18n` directory
```

The [`i18n` directory](packages/studio-base/src/i18n) contains translated (localized) strings for all languages supported by Foxglove Studio.

> Translated strings are organized into _namespaces_ — for example, [`i18n/[language]/preferences.ts`](packages/studio-base/src/i18n/en/preferences.ts) contains localization keys and strings for the app's Preferences window.

Reviewer: ```suggestion
Translated strings are organized into _namespaces_ — e.g. [`i18n/[language]/preferences.ts`](packages/studio-base/src/i18n/en/preferences.ts) contains translations for the app's Preferences tab.
```

### Use `useTranslation()` and `t()` to access translated strings

1. Call the [<code>useTranslation(<i>namespace</i>)</code> hook](https://react.i18next.com/latest/usetranslation-hook) inside a React component to access strings in a given namespace. The hook returns a function called `t`.

> 2. Call the `t` function to get the translated version of a string.

Reviewer: ```suggestion
2. Call the `t` function to get the translation of a string.
```

For example:

```ts
const { t } = useTranslation("myComponent");
return <p>{t("hello")}</p>;
```

> ### Adding localization support to a component

Reviewer: ```suggestion
### Add localization support to a component
```

> 1. Move English strings out of the component code, and into the `i18n` folder. Use a new namespace for logical groups of components or app screens.

Reviewer: ```suggestion
1. Move English strings out of the component code, and into the `i18n` folder. Use a new namespace for logical groups of components or app views.
```

> 2. Replace strings hard-coded in source code with calls to the `t()` function. Use `camelCase` for keys.

Reviewer: ```suggestion
2. Replace strings hard-coded in source code with calls to the `t()` function. Use `camelCase` for new localization keys.
```

<table><tr><th>Before</th><th>After</th></tr><tr><td>

```ts
function MyComponent() {
  return <p>Hello!</p>;
}
```

</td><td>

```ts
function MyComponent() {
  const { t } = useTranslation("myComponent");
  return <p>{t("hello")}</p>;
}
```

```ts
// i18n/en/myComponent.ts
export default {
  hello: "Hello!",
};
```

</td></tr></table>

### Complete example

```ts
// MyComponent.ts

import { useTranslation } from "react-i18next";

function MyComponent(props: Props): JSX.Element {
  const { t } = useTranslation("myComponent");

  return <p>{t("hello")}</p>;
}
```

```ts
// i18n/index.ts
export const translations = {
  en: {
    ...,
    myComponent: enMyComponent,
  },
  zh: {
    ...,
    myComponent: zhMyComponent,
  },
};
```

```ts
// i18n/en/myComponent.ts
export default {
  hello: "Hello!",
};

// i18n/en/index.ts
export { default as enMyComponent } from "./myComponent";
```

```ts
// i18n/zh/myComponent.ts
export default {
  hello: "你好！",
};

// i18n/zh/index.ts
export { default as zhMyComponent } from "./myComponent";
```

Result:

| English         | Chinese         |
| --------------- | --------------- |

> | `<p>Hello!</p>` | `<p>你好！</p>` |

Reviewer: ```suggestion
| `<p>Hello!</p>` | `<p>你好</p>` |
```

> | `<p>Hello!</p>` | `<p>你好！</p>` |

Reviewer: #nit - is that last space in the "Chinese" column supposed to be there?

Author: there's no space, it's U+FF01 FULLWIDTH EXCLAMATION MARK
